pageserver/src/tenant/ephemeral_file/page_caching.rs

```diff
@@ -96,10 +103,116 @@ impl Drop for RW {
                 // not found files might also be related to https://github.com/neondatabase/neon/issues/2442
                 error!(
                     "could not remove ephemeral file '{}': {}",
-                    self.rw.as_inner_virtual_file().path,
+                    self.rw.as_writer().file.path,
                     e
                 );
             }
         }
     }
 }
+
+struct PreWarmingWriter {
+    nwritten_blocks: u32,
+    page_cache_file_id: page_cache::FileId,
+    file: VirtualFile,
+}
+
+impl PreWarmingWriter {
+    fn new(page_cache_file_id: page_cache::FileId, file: VirtualFile) -> Self {
+        Self {
+            nwritten_blocks: 0,
+            page_cache_file_id,
+            file,
+        }
+    }
+}
+
+impl crate::virtual_file::owned_buffers_io::write::OwnedAsyncWriter for PreWarmingWriter {
+    async fn write_all<
+        B: tokio_epoll_uring::BoundedBuf<Buf = Buf>,
+        Buf: tokio_epoll_uring::IoBuf + Send,
+    >(
+        &mut self,
+        buf: B,
+    ) -> std::io::Result<(usize, B::Buf)> {
+        let buf = buf.slice(..); // the resulting Slice derefs to &[u8] of [0..buf.bytes_init()]
+        let saved_bounds = buf.bounds(); // save for reconstructing buf from iobuf after the IO is done
```

Comment:
The resulting Slice does indeed deref to `&[u8] of [0..buf.bytes_init()]`, but the bounds will be `0..buf.bytes_total()`. That's fine for `zero_padded::Buffer` (`PreWarmingWriter` is only used with this buf type), but I'm not sure what happens if the buffer were to change.

Reply:
In C++ I'd add a `static_assert` to ensure that the buffer type of `PrewarmingWriter` is `zero_padded::Buffer`, but I'm not sure what the rust equivalent is.

Reply:
Actually, I think `0..buf.bytes_total()` is what you want for reconstructing the buffer.

Reply:
Maybe the wording is wrong.

What I want to achieve here is the following: after the IO returns, I need the same &[u8] slice as what was effectively written to the VirtualFile, such that I can pre-warm the page cache with it.

Will try to improve the wording in the comments.

Reply:
WDYT of d928ad67a390351c4d6b0d45ef88b57765813eaa

Reply:
Yeah, makes sense. This thread was more of a train of thought while reviewing as opposed to something actionable.